### PR TITLE
gravel/glass: user management

### PR DESCRIPTION
**Pull request by votdev**
The default credentials set up during deployment are: admin/aquarium

Peek 2021-06-25 17-27

Peek 2021-06-25 17-40

Fixes: https://github.com/aquarist-labs/aquarium/issues/421
Fixes: https://github.com/aquarist-labs/aquarium/issues/23
Fixes: https://github.com/aquarist-labs/aquarium/issues/26

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References issues, create if required
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test tumbleweed`
- `jenkins run tumbleweed`
- `jenkins test leap`
- `jenkins run leap`

</details>

### CODE REVIEW

File: src/aquarium.py
Comment (jecluis) [inline src/aquarium.py:51]: This should be conditional, somehow. Maybe passing an environment variable or something.

File: src/gravel/api/auth.py
Comment (jecluis) [inline src/gravel/api/auth.py:37]: specify the return type

File: src/gravel/api/auth.py
Comment (jecluis) [inline src/gravel/api/auth.py:46]: s/login/logout/

File: src/gravel/api/user.py
Comment (jecluis) [inline src/gravel/api/user.py:32]: specify return type

File: src/gravel/api/user.py
Comment (jecluis) [inline src/gravel/api/user.py:37]: specify return type

File: src/gravel/api/user.py
Comment (jecluis) [inline src/gravel/api/user.py:42]: return type

File: src/gravel/api/user.py
Comment (jecluis) [inline src/gravel/api/user.py:47]: return type

File: src/gravel/controllers/user.py
Comment (jecluis) [inline src/gravel/controllers/user.py:32]: if we're setting the default as `False`, I don't see a reason to keep this as `Optional`. If the field is not specified, it will take the default value.

File: src/gravel/controllers/user.py
Comment (jecluis) [inline src/gravel/controllers/user.py:49]: This could be a `parse_raw()` instead of having to do a `json.loads()` and a `parse_obj()`, I think.

File: src/gravel/controllers/user.py
Comment (jecluis) [inline src/gravel/controllers/user.py:59]: same here, I think this could be a `parse_raw()`.

File: src/gravel/api/user.py
Comment (jecluis) [inline src/gravel/api/user.py:43]: This might be returning "None", which is something we probably want to check and convey something more descriptive to the user, even if by raising an exception with a status code.

**Review by jecluis — CHANGES_REQUESTED** [review]
few things of note, and we need tests for this.

File: src/aquarium.py
Comment (jecluis) [reply src/aquarium.py:51]: Also, this is a dependency that is not being added. Which I support; I don't think adding this dependency is something we want. So maybe do this if the dependency exists, like in a `try` block of the sorts of

```
try:
  import pydevd_pycharm
  pydevd_pycharm.settrace(...)
except ModuleNotFound:
  pass
```

Comment (votdev) [reply src/aquarium.py:54]: This will be removed in the final PR.

File: src/gravel/controllers/user.py
Comment (votdev) [reply src/gravel/controllers/user.py:49]: > This could be a parse_raw() instead of having to do a json.loads() and a parse_obj(), I think.

Right, according to https://pydantic-docs.helpmanual.io/usage/models/#helper-functions this is correct.

**Comment by votdev** [comment]
jenkins run tumbleweed

**Comment by votdev** [comment]
jenkins run leap

**Comment by votdev** [comment]
jenkins test tumbleweed

**Comment by votdev** [comment]
jenkins test leap

**Comment by votdev** [comment]
jenkins test tumbleweed

**Comment by votdev** [comment]
> Not sure if it's part of this PR, but the red notification which pops up after a user has logged out is not something I'd prefer to show a user on successful logout. Due to the red color it looks like an error and the message seems to be more helpful to developers than to users. Maybe it's better to omit it?

Fixed

File: src/glass/src/app/shared/services/auth-storage.service.ts
Comment (jecluis) [inline src/glass/src/app/shared/services/auth-storage.service.ts:25]: where is `localStorage` coming from?

File: src/gravel/api/auth.py
Comment (jecluis) [inline src/gravel/api/auth.py:20]: Please limit this at 80 cols, and split  them into lines in the format
```
from gravel.controllers.auth import (
  jwt_auth_scheme,
  ...
)
```

File: src/gravel/api/auth.py
Comment (jecluis) [inline src/gravel/api/auth.py:38]: I'd prefer the return to be moved to the next line and properly limit the length at 80 cols. Something like

```
async def login(request: Request, response: Response,
                form_data: OAuth2PasswordRequestForm = Depends()
) -> LoginReplyModel:
```

File: src/gravel/controllers/auth.py
Comment (jecluis) [inline src/gravel/controllers/auth.py:81]: mind adding comments explaining what each one of these things are?

File: src/gravel/controllers/auth.py
Comment (jecluis) [inline src/gravel/controllers/auth.py:169]: The API is leaking here by passing a `Request` to the controller. It would be preferable to pass what this class actually requires (seems it's the `GlobalState`?) instead of a fastapi request.

File: src/gravel/controllers/auth.py
Comment (jecluis) [inline src/gravel/controllers/auth.py:186]: Raise a controller specific exception, and translate that to whatever is the appropriate HTTPException in the API controller.

File: src/gravel/controllers/auth.py
Comment (jecluis) [inline src/gravel/controllers/auth.py:194]: same here, and below.

**Review by jecluis — CHANGES_REQUESTED** [review]
Few comments. In terms of style, mostly have issues with line length and line breaks (or lack of), but given we are not enforcing them I'm not going to hold the PR on that account.

We are leaking a lot of API state and structures to the core controllers, which should not happen, as it will make unit testing harder, and from an architectural point of view it makes little sense IMO.

Also, we're lacking unit tests for the added structures. We'll need tests.

File: src/aquarium.py
Comment (votdev) [reply src/aquarium.py:16]: Right, strange, my IDE suggests to remove it.

File: src/glass/src/app/shared/services/auth-storage.service.ts
Comment (votdev) [reply src/glass/src/app/shared/services/auth-storage.service.ts:25]: It's global available. See https://developer.mozilla.org/de/docs/Web/API/Window/localStorage

Comment (votdev) [reply src/glass/src/app/shared/services/auth-storage.service.ts:25]: You can also use `window.localStorage`.

File: src/aquarium.py
Comment (votdev) [reply src/aquarium.py:16]: Do we really need `import logging.config`?

File: src/gravel/api/auth.py
Comment (votdev) [reply src/gravel/api/auth.py:38]: > I'd prefer the return to be moved to the next line and properly limit the length at 80 cols. Something like
> 
> ```
> async def login(request: Request, response: Response,
>                 form_data: OAuth2PasswordRequestForm = Depends()
> ) -> LoginReplyModel:
> ```

My IDE does not like that and shows a PEP8 error. Instead we should format it this way:
```
async def login(
        request: Request, response: Response,
        form_data: OAuth2PasswordRequestForm = Depends()
) -> LoginReplyModel:
```

File: src/aquarium.py
Comment (jecluis) [reply src/aquarium.py:16]: `logging.config`, I don't think so.

File: src/gravel/api/auth.py
Comment (jecluis) [reply src/gravel/api/auth.py:38]: I most definitely agree. I just didn't want to go and be too annoying :)

File: src/gravel/controllers/auth.py
Comment (votdev) [reply src/gravel/controllers/auth.py:169]: The method signature is given by the derived class [OAuth2PasswordBearer](https://github.com/tiangolo/fastapi/blob/master/fastapi/security/oauth2.py#L152)

Can we relocate this class to somewhere else in our source tree where we do not pollute our `controller` rules?

File: src/gravel/controllers/auth.py
Comment (votdev) [reply src/gravel/controllers/auth.py:186]: That's not possible, an instance of this class is injected in every API endpoint and is processed by FastAPI internally. No way to catch and map it.

File: src/gravel/controllers/auth.py
Comment (jecluis) [reply src/gravel/controllers/auth.py:169]: If this is bound to the API, then it should probably live under `api/auth.py`.

File: src/gravel/controllers/auth.py
Comment (jecluis) [reply src/gravel/controllers/auth.py:186]: By moving this to the `api/auth.py` file, we get rid of this annoyance and we raising HTTPExceptions still make sense IMO.

**Comment by s0nea** [comment]
I just setup a testing environment for the PR. While doing this two questions came into my mind:

-  Would it be sensible to print the default username/password by e.g. the `run_aquarium.sh` script?
or
- If it's possible: is there a plan to define the default credentials while running the install wizard?

**Comment by votdev** [comment]
>     * If it's possible: is there a plan to define the default credentials while running the install wizard?

As far as i remember we came to the conclusion to do not request the username and password during the installation.

**Comment by jecluis** [comment]
> I just setup a testing environment for the PR. While doing this two questions came into my mind:
> 
>     * Would it be sensible to print the default username/password by e.g. the `run_aquarium.sh` script?

I think this makes little sense given the user won't be able to see the console.

>     * If it's possible: is there a plan to define the default credentials while running the install wizard?

And I'd think that's a "yes", and a "must have". The user should be able to configure a username and a password during initial deployment, and even be able to provide an ssh pubkey to be able to ssh into the nodes (maybe?).

**Comment by votdev** [comment]
> > I just setup a testing environment for the PR. While doing this two questions came into my mind:
> > ```
> > * Would it be sensible to print the default username/password by e.g. the `run_aquarium.sh` script?
> > ```
> 
> I think this makes little sense given the user won't be able to see the console.
> 
> > ```
> > * If it's possible: is there a plan to define the default credentials while running the install wizard?
> > ```
> 
> And I'd think that's a "yes", and a "must have". The user should be able to configure a username and a password during initial deployment, and even be able to provide an ssh pubkey to be able to ssh into the nodes (maybe?).

:-) Two different opinions. 

OK, i can add a username and password page to the `Create` wizard. Th eSSH pubkey should be done in a separate issue/PR.

**Comment by s0nea** [comment]
> > Would it be sensible to print the default username/password by e.g. the run_aquarium.sh script?
> 
> The displayed information will be obsolete as soon as the password has been changed or the user jas been replaced by another one. Because of that it doesn't make sense to display it.

I thought about it as well. But given the fact the script is mostly used for development and testing it's worth it in my opinion.

**Comment by s0nea** [comment]
> > I just setup a testing environment for the PR. While doing this two questions came into my mind:
> > ```
> > * Would it be sensible to print the default username/password by e.g. the `run_aquarium.sh` script?
> > ```
> 
> I think this makes little sense given the user won't be able to see the console.

It was just an idea because I setup the environment and then I had to look up the credentials in the code.

**Comment by jecluis** [comment]
> OK, i can add a username and password page to the `Create` wizard. The SSH pubkey should be done in a separate issue/PR.

I think both can/should be a different PR. We might not want to hold this one right now. Also, this one still needs tests to be merged, even if basic ones.

File: src/glass/src/app/pages/login-page/login-page.component.html
Comment (s0nea) [inline src/glass/src/app/pages/login-page/login-page.component.html:11]: Would it be possible to call `doLogin()` by hitting 'enter' after entering the password? Just a minor thing. It shouldn't block the PR from being merged and can be addressed in a different issue if you prefer.

File: src/glass/src/app/pages/users-page/users-page.component.ts
Comment (s0nea) [inline src/glass/src/app/pages/users-page/users-page.component.ts:179]: Can also be addressed in a separate issue: I do have the feeling we should prevent the logged-in user from deleting himself.

File: src/glass/src/app/pages/login-page/login-page.component.html
Comment (votdev) [reply src/glass/src/app/pages/login-page/login-page.component.html:11]: No, that's not possible because the submit button is not part of the form. If this is required, we need to refactor the form implementation.

Comment (votdev) [reply src/glass/src/app/pages/login-page/login-page.component.html:11]: I'll check that, didn't find the reason why it does not work till now.

File: src/glass/src/app/pages/users-page/users-page.component.ts
Comment (votdev) [reply src/glass/src/app/pages/users-page/users-page.component.ts:179]: The API endpoint will check that.

File: src/glass/src/app/pages/login-page/login-page.component.html
Comment (votdev) [reply src/glass/src/app/pages/login-page/login-page.component.html:12]: Fixed

Comment (s0nea) [reply src/glass/src/app/pages/login-page/login-page.component.html:12]: Nice :partying_face: Thank you!

**Comment by votdev** [comment]
jenkins test tumbleweed

File: src/glass/src/app/pages/users-page/users-page.component.ts
Comment (s0nea) [reply src/glass/src/app/pages/users-page/users-page.component.ts:179]: :+1: I think it would be great if the user **can't** disable himself in addition.

**Comment by s0nea** [comment]
Somehow I can't re-enable a deactivated user:

https://user-images.githubusercontent.com/8761082/124268524-e7973f80-db39-11eb-9e89-a2c2779e98a5.mp4

Any ideas?

**Comment by votdev** [comment]
> Somehow I can't re-enable a deactivated user:
> aqua.mp4
> 
> Any ideas?

Fixed

**Comment by s0nea** [comment]
jenkins test leap

File: src/glass/src/app/shared/components/submit-button/submit-button.component.ts
Comment (p-se) [inline src/glass/src/app/shared/components/submit-button/submit-button.component.ts:26]: incomplete comment?

Comment (votdev) [reply src/glass/src/app/shared/components/submit-button/submit-button.component.ts]: Fixed

**Comment by votdev** [comment]
jenkins test tumbleweed

**Review by s0nea — APPROVED** [review]
The `glass` part LGTM!

I suggest to wait for @jecluis's final review before merging the PR.

**Review by jecluis — APPROVED** [review]

**Pull request merged**
